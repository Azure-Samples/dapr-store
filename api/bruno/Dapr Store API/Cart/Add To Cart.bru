meta {
  name: Add To Cart
  type: http
  seq: 1
}

put {
  url: {{endpoint}}/cart/method/setProduct/{{userId}}/prd001/6
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isNotEmpty
}
