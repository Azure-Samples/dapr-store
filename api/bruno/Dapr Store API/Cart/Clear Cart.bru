meta {
  name: Clear Cart
  type: http
  seq: 4
}

put {
  url: {{endpoint}}/cart/method/clear/{{userId}}
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isEmpty
}
