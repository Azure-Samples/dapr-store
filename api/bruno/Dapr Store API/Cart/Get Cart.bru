meta {
  name: Get Cart
  type: http
  seq: 2
}

get {
  url: {{endpoint}}/cart/method/get/{{userId}}
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isNotEmpty
}
