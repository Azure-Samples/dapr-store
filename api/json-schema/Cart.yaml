$schema: https://json-schema.org/draft/2020-12/schema
$id: Cart.yaml
type: object
properties:
  products:
    $ref: "#/$defs/RecordInt32"
  forUserId:
    type: string
required:
  - products
  - forUserId
$defs:
  RecordInt32:
    type: object
    properties: {}
    unevaluatedProperties:
      type: integer
      minimum: -2147483648
      maximum: 2147483647
