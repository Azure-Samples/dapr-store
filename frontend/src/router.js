import VueRouter from 'vue-router'

import ProductCatalog from './components/ProductCatalog'
import ProductOffers from './components/ProductOffers'
import Home from './components/Home'

const router = new VueRouter({
  routes: [
    {
      path: '/catalog',
      component: ProductCatalog
    },
    {
      path: '/offers',
      component: ProductOffers
    },
    {
      path: '/',
      component: Home
    }
  ]
})

export default router