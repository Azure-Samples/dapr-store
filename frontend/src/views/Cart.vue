<template>
  <div>
    <h1><fa icon="shopping-cart" /> &nbsp; Shopping Cart</h1><br>
    <b-table striped hover :items="user.cart" :fields="fields" />
    <b-button :disabled="user.cart.length == 0" variant="primary" size="lg">
      <fa icon="shopping-basket" /> &nbsp; CHECKOUT
    </b-button>
  </div>
</template>

<script>
import { userProfile } from '../main'

export default {
  name: 'Cart',

  data() {
    return {
      fields: ['name', 'cost', 'description'],
      user: userProfile
    }
  },
}
</script>

<style>

</style>