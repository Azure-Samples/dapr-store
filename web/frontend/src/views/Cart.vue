<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - Cart contents view
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <h1>
      <fa icon="shopping-cart" /> &nbsp; Shopping Cart
    </h1>
    <br>
    <error-box :error="error" />

    <div v-if="!cart && !error" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <b-alert v-if="newOrder" show variant="success">
      <h4>Order accepted! ðŸ˜„</h4>
      <b-link :to="`order/`+newOrder.id">
        <div>
          Your order for <b>Â£{{ newOrder.amount }}</b> has been accepted, the order ID is: <b>{{ newOrder.id }}</b><br>
          Check your account for progress on your order(s)
        </div>
      </b-link>
    </b-alert>

    <b-alert v-if="cart && Object.keys(cart.products).length==0" show variant="dark">
      Your cart is empty! <br><br><b-link href="/catalog">
        Go shopping!
      </b-link>
    </b-alert>

    <div v-if="cart">
      <h2>Cart total: Â£{{ total.toFixed(2) }} </h2>
      <b-card v-for="product of cartProducts" :key="product.id" class="m-3 p-1" header-bg-variant="primary" header-text-variant="white">
        <template v-slot:header>
          <span>{{ product.name }}</span>
        </template>
        <h2>Count:</h2> <input :value="cart.products[product.id]" type="text" readonly>

        <b-button class="ml-5 mr-3" :disabled="!cart || cart.products.length == 0" variant="warning" size="lg" @click="modifyProductAmmount(product.id, -1)">
          <fa icon="minus-circle" />
        </b-button>
        <b-button :disabled="!cart || cart.products.length == 0" variant="success" size="lg" @click="modifyProductAmmount(product.id, 1)">
          <fa icon="plus-circle" />
        </b-button>
        <img :src="product.image" class="thumb">
      </b-card>
    </div>

    <b-button :disabled="!cart || cartProducts.length == 0" variant="primary" size="lg" @click="submitOrder">
      <fa icon="shopping-basket" /> &nbsp; CHECKOUT
    </b-button>
    &nbsp;
    <b-button :disabled="!cart || cartProducts.length == 0" variant="warning" size="lg" class="float-right" @click="clearCart">
      <fa icon="trash-alt" /> &nbsp; EMPTY CART
    </b-button>
  </div>
</template>

<script>
import ErrorBox from '../components/ErrorBox'
import api from '../services/api'
import auth from '../services/auth'

export default {
  name: 'Cart',

  components: {
    'error-box': ErrorBox
  },

  data() {
    return {
      error: null,
      newOrder: null,
      cart: null,
      cartProducts: [],
      user: null
    }
  },

  computed: {
    total() {
      let tot = 0
      for (let product of this.cartProducts) {
        let count = this.cart.products[product.id]
        tot += (count * product.cost)
      }

      return tot
    }
  },

  async mounted() {
    try {
      this.user = auth.user()
      if (!this.user) { return }

      let resp = await api.cartGet(this.user.userName)
      if (resp) {
        this.cart = resp
        this.cartProducts = []
        for (let productId in this.cart.products) {
          // Do this async helps speed it up when running locally, due to Dapr bug
          api.productGet(productId)
            .then((resp) => {
              this.cartProducts.push(resp)
            })
        }
      }
    } catch (err) {
      this.error = err
    }
  },

  methods: {
    async submitOrder() {
      try {
        this.newOrder = await api.cartSubmit(this.user.userName)
        this.cart = await api.cartClear(this.user.userName)
        this.cartProducts = []
      } catch (err) {
        this.error = err
      }
    },

    async clearCart() {
      try {
        let resp = await api.cartClear(this.user.userName)
        this.cart = resp
        this.cartProducts = []
      } catch (err) {
        this.error = err
      }
    },

    async modifyProductAmmount(productId, amount) {
      try {
        this.cart = await api.cartAddAmount(this.user.userName, productId, amount)

        // Fiddly nonsense to remove from cartProducts if removed from products.cart
        // Check if productId is removed from cart object, then recreate cartProducts array
        if (!Object.prototype.hasOwnProperty.call(this.cart.products, productId)) {
          this.cartProducts = this.cartProducts.filter((p) => p.id != productId)
        }
      } catch (err) {
        this.error = err
      }
    }
  }
}
</script>

<style scoped>
  input[type=text] {
    border: none;
    width: 4rem;
    text-align: center;
  }
  .thumb {
    width: 80px;
    float: right;
  }
</style>