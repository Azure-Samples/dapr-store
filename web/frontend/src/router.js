// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - client side routes
// ----------------------------------------------------------------------------

import { createRouter, createWebHistory } from 'vue-router'

import auth from './services/auth'

import ProductCatalog from './views/ProductCatalog.vue'
import ProductOffers from './views/ProductOffers.vue'
import ProductSearch from './views/ProductSearch.vue'
import ProductSingle from './views/ProductSingle.vue'
import ViewOrder from './views/ViewOrder.vue'
import Login from './views/Login.vue'
import Account from './views/Account.vue'
import Home from './views/Home.vue'
import About from './views/About.vue'
import Cart from './views/Cart.vue'

const router = createRouter({
  history: createWebHistory(),
  routes: [
    {
      name: 'home',
      path: '/',
      component: Home
    },
    {
      name: 'login',
      path: '/login',
      component: Login
    },
    {
      name: 'account',
      path: '/account',
      component: Account,
      beforeEnter: signedInCheck
    },
    {
      name: 'cart',
      path: '/cart',
      component: Cart,
      beforeEnter: signedInCheck
    },
    {
      name: 'about',
      path: '/about',
      component: About
    },
    {
      name: 'search',
      path: '/search/:query',
      component: ProductSearch
    },
    {
      name: 'catalog',
      path: '/catalog',
      component: ProductCatalog
    },
    {
      name: 'offers',
      path: '/offers',
      component: ProductOffers
    },
    {
      name: 'single-product',
      path: '/product/:id',
      component: ProductSingle
    },
    {
      name: 'view-order',
      path: '/order/:id',
      component: ViewOrder
      //beforeEnter: signedInCheck
    }
  ]
})

function signedInCheck(to, from, next) {
  const user = auth.user()
  if (!user || !user.username) {
    next('/login')
  } else {
    next()
  }
}

export default router
