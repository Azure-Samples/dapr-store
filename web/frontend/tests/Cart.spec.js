import { describe, it, vi, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import router from '@/router'
import { flushPromises } from '@vue/test-utils'

import Cart from '@/views/Cart.vue'

vi.mock('@/services/api')
vi.mock('@/services/auth')

describe('Cart.vue', () => {
  it('renders cart contents', async () => {
    const wrapper = mount(Cart, {
      global: {
        plugins: [router]
      }
    })

    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
