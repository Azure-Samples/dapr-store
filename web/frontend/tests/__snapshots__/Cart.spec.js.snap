// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart.vue > renders cart contents 1`] = `
"<div data-v-7d4f60e4="">
  <h1 data-v-7d4f60e4=""><i data-v-7d4f60e4="" class="fa-solid fa-cart-shopping"></i> &nbsp; Shopping Cart</h1><br data-v-7d4f60e4="">
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
  <div data-v-7d4f60e4="">
    <h2 data-v-7d4f60e4="">Cart total: £91.10</h2>
    <div data-v-7d4f60e4="" class="card m-3 p-1" header-bg-variant="primary" header-text-variant="white">
      <div data-v-7d4f60e4="" class="card-header">Top Hat (6″)</div>
      <div data-v-7d4f60e4="" class="card-body">
        <h2 data-v-7d4f60e4="">Count:</h2><input data-v-7d4f60e4="" type="text" readonly="" class="m-3" value="2"><button data-v-7d4f60e4="" class="btn btn-warning p-2 px-3 mx-3"><i data-v-7d4f60e4="" class="fa-solid fa-circle-minus"></i></button><button data-v-7d4f60e4="" class="btn btn-success p-2 px-3"><i data-v-7d4f60e4="" class="fa-solid fa-circle-plus"></i></button><img data-v-7d4f60e4="" src="/img/catalog/1.jpg" class="thumb">
      </div>
    </div>
    <div data-v-7d4f60e4="" class="card m-3 p-1" header-bg-variant="primary" header-text-variant="white">
      <div data-v-7d4f60e4="" class="card-header">Paisley Cravat Ascot Tie</div>
      <div data-v-7d4f60e4="" class="card-body">
        <h2 data-v-7d4f60e4="">Count:</h2><input data-v-7d4f60e4="" type="text" readonly="" class="m-3" value="1"><button data-v-7d4f60e4="" class="btn btn-warning p-2 px-3 mx-3"><i data-v-7d4f60e4="" class="fa-solid fa-circle-minus"></i></button><button data-v-7d4f60e4="" class="btn btn-success p-2 px-3"><i data-v-7d4f60e4="" class="fa-solid fa-circle-plus"></i></button><img data-v-7d4f60e4="" src="/img/catalog/3.jpg" class="thumb">
      </div>
    </div>
  </div>
  <div data-v-7d4f60e4="" class="d-flex justify-content-between mt-4"><button data-v-7d4f60e4="" class="btn btn-lg btn-primary"><i data-v-7d4f60e4="" class="fa-solid fa-basket-shopping"></i> &nbsp; CHECKOUT </button><button data-v-7d4f60e4="" class="btn btn-lg btn-warning"><i data-v-7d4f60e4="" class="fa-solid fa-trash-can"></i> &nbsp; EMPTY CART </button></div>
</div>"
`;
